04.ls/ls.rb

```diff
@@ -0,0 +1,30 @@
+#! /usr/bin/env ruby
+# frozen_string_literal: true
+
+COLUMNS = 3
+COLUMNS_DISTANCE = 5
+
+def main
+  longest_file_name = file_names.max_by(&:size)
+  file_name_width = longest_file_name.size + COLUMNS_DISTANCE
+  build_output_file_names.each do |row|
+    row.each do |file_name|
+      print file_name.to_s.ljust(file_name_width)
+    end
+    puts
+  end
+end
+
+def build_output_file_names
+  rows = file_names.size.ceildiv(COLUMNS)
+  additional_blanks = COLUMNS * rows - file_names.size
+  full_file_names = file_names + Array.new(additional_blanks)
+  sliced_file_names = full_file_names.each_slice(rows).to_a
+  sliced_file_names.transpose
+end
+
+def file_names
```

Comment:
結果が同じになるにもかかわらず、このメソッドが何度も呼ばれていますね。
最近のCPUはとても速いので実感がすることはまれですが、それでも何らかの処理を行うというのはゼロコストではありません。
たとえば「ここで処理が走ってますよ」ということを意識できるように、sleepを挟んでみてください。

```suggestion
def file_names
  sleep 2
```

理想はこのsleepが1回だけ呼ばれる、つまり2秒でプログラムが完了することです。

Reply:
`file_names`メソッドに`sleep 2`を加えてプロクラムを実行しました。出力されるのに時間がかかりました。
```ruby
def file_names
  sleep 2
  Dir['*'].sort_by(&:downcase)
end
```
Benchmarkモジュールというのを見つけたので、`build_output_file_names`の実行時間も調べて、`file_names`メソッドが3回呼ばれていることが推測できました。
```ruby
Benchmark.bm do |x|
  x.report('buiid') {build_output_file_names}
end

# =>            user     system      total        real
# => buiid  0.001309   0.001340   0.002649 (  6.017689)
```
`build_output_file_names`メソッドに`binding.break`を置いて、実際に`file_names`メソッドの中に3回入っていくのを確認しました。
```ruby
def build_output_file_names
  binding.break
  rows =  file_names.size.ceildiv(COLUMNS)
  # 省略
  sliced_file_names.transpose
end
```

Reply:
`file_names`メソッドの返り値を変数に代入するように修正しました。
`sleep 2`を加えると、出力されるまで時間がかかりました。
修正後は2秒でメソッドを抜けるようになりました。
```ruby
# メソッドの実行にかかる時間を調べました。


# 省略

def file_names
  sleep 2
  Dir['*'].sort_by(&:downcase)
end

main

Benchmark.bm do |x|
  x.report('buiid') {build_output_file_names}
end

# 修正前
# =>            user     system      total        real
# => buiid  0.001736   0.001895   0.003631 (  6.014195)

# 修正後
# =>            user     system      total        real
# => buiid  0.000326   0.000707   0.001033 (  2.006036)
```

Reply:
ちゃんとベンチマークを取ってるのはいいですね  👍 
ただ、対象が`build_output_file_names`だけになっています。
着目すべきはこのメソッドだけでなく、プログラム全体です。
なのでmainメソッドを対象にしてください。
もしくはtimeコマンドを使うのもありです。

```
$ time ls.rb
```

Reply:
`file_names`の結果を引数で`main`と`build_output_file_names`に渡すように修正しました。
`file_names`に`sleep 2`を挟むと、実行時間が2秒強になりました。
```
$ time ls.rb
1fx7r4ZrFYlUY     BpeUhiLtx         qSBMpfgAr         
4Rd7              bPu               SWPLR5AQf         
8lEYP             Engayr            TqBPe             
9mzun4lC          fQmI3Yt           y45E87ehkEL       
BImEQ2            podsLfU                             
ls.rb  0.04s user 0.03s system 3% cpu 2.180 total
```